File: practice/algorithms/implementation/absolute_permutation/absolute_permutation.cpp
#include <set>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

int main()
{
    int t, n, k;
    cin >> t;
    while (t--)
    {
        cin >> n >> k;
        vector<int> perm(n), check;
        set<int> used;
        
        for (int pos = 1; pos < n + 1; ++pos)
        {
            int d = pos - k;
            if (d > 0 && (used.find(d) == used.end()))
            {
                perm[pos - 1] = d;
                used.insert(d);
            }
            else
            {
                perm[pos - 1] = pos + k;
                used.insert(pos + k);
            }
        }
        
        check = perm;
        sort(check.begin(), check.end());
        bool is_valid = true;
        for (int pos = 0; pos < n; ++pos)
        {
            if (check[pos] != (pos + 1))
            {
                is_valid = false;
                break;
            }
        }
        
        if (is_valid)
        {
            for (int pos = 0; pos < n; ++pos)
            {
                cout << perm[pos] << " ";
            }
            cout << endl;
        }
        else
        {
            cout << -1 << endl;
        }
    }
    return 0;
}

File: practice/algorithms/implementation/absolute_permutation/absolute_permutation.hs
module Main where
import Control.Monad (forM_)
import qualified Data.Set as Set
import Data.List

generateNewValue :: Int -> Int -> Set.Set Int -> Int
generateNewValue pos k used
    | d > 0 && Set.member d used == False = d
    | otherwise                           = pos + k
    where
        d = pos - k

solve :: Int -> Int -> ([Int], Set.Set Int)
solve n k = foldr (\pos acc -> process pos k acc) ([], Set.empty) [n,n-1..1]
    where
        process pos k res = (value : (fst res), Set.insert value (snd res))
            where value = generateNewValue pos k (snd res)

main :: IO ()
main = do
    t <- readLn :: IO Int
    forM_ [1..t] $ \_  -> do
        params <- getLine
        let [n, k] = map read $ words params :: [Int]
            res_p  = reverse $ fst $ solve n k
        if sort res_p == [1..n]
            then putStrLn $ unwords $ map show res_p
            else print (-1)

